pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +181,26 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, repo_id, file_obj):
+        upload_fts = []
+
+        if isinstance(file_obj, str):
+            file_obj = open(file_obj, "rb")
+
+        checksum = hashlib.sha256()
+        size, offset = 0, 0
+        data = file_obj.read(self._CHUNK_SIZE)
+        LOG.info("Uploading %s to repo %s", file_obj, repo_id)
+        while data:
```

Comment:
It seems like it's still doing too much up front in a blocking style.

Consider for example if you want to upload a 4GB ISO.  This will read the entire file into memory and enqueue 4,000 futures each holding a reference to a 1MB chunk.  So this method will have consumed 4GB of memory before returning any future.

I don't think it's right, it means that you can't upload any file with a size larger than the amount of memory you can allocate.  I don't think we should accept such a limit.

I think that you need to commit to making this either fully non-blocking (e.g. one future reads one chunk, next future reads another chunk ...), or if it's too complicated then make it fully blocking (a loop where you read a chunk, upload it, wait for that upload to finish, read next chunk, upload it, wait for that to finish ... etc).